### ISSUE TITLE
Onedrive cloud provider?

### ISSUE THREAD

**Issue by VGuerreiro** [johannesjo/super-productivity#2109]
First of all, I want to thank you for this amazing app!

Are there any plans to add OneDrive as a cloud provider for the Sync&Export functionality? 
There are options for DropBox and GoogleDrive but I can't figure out how to set it up with OneDrive.

**Comment by Evynly** [comment]
@VGuerreiro You should be able to connect to OneDrive through WebDAV - check out [this guide](https://winscp.net/eng/docs/guide_microsoft_onedrive) (while it isn't specific to Super-Productivity, the steps performed to set up the connection should be similar)

**Comment by ahdung** [comment]
+1

**Comment by RokeJulianLockhart** [comment]
> https://github.com/johannesjo/super-productivity/issues/2109#event-12217314442

@johannesjo, worth removing https://github.com/johannesjo/super-productivity/issues?q=label%3AStale again?

**Comment by johannesjo** [comment]
@RokeJulianLockhart yes! :)

**Comment by Joyfulbeekeeper** [comment]
Still no direct OneDrive, Yandex and Mail.ru Cloud sync in 2025

**Comment by XanatosX** [comment]
Please reopen, would be nice Rob have a native OneDrive provider.

**Comment by beedell-roke** [comment]
> https://github.com/johannesjo/super-productivity/issues/2109#issue-1260894519

@VGuerreiro, why not utilise [`Microsoft.OneDrive`](https://github.com/microsoft/winget-pkgs/blob/f93572949ffc261ab14d4f782ba6db0a44204b46/manifests/m/Microsoft/OneDrive/25.070.0413.0001/Microsoft.OneDrive.installer.yaml#L42C17-L42C97:~:text=InstallerUrl:-,https://oneclient.sfx.ms/Win/Installers/25.070.0413.0001/amd64/OneDriveSetup.exe,-InstallerSha256:%20CFABD20FA85F874CB93E14C34BCDCA8B1DAB82EC934B818601B698F3CBA86717) or [`github.com/abraunegg/onedrive`](https://github.com/abraunegg/onedrive/blob/810197cc05d81207fbfda86ba8fd1d248d6c3943/readme.md#onedrive-client-for-linux:~:text=Raw-,OneDrive%20Client%20for%20Linux,-Version%20Release%20Date) (so via the filesystem), in the stead of an application-specific integration? I don't see the value to this.

**Comment by XanatosX** [comment]
This would not work on a mobile device like DropBox does. At leata I cannot imagen how this should work.

Did I miss something?

**Comment by beedell-roke** [comment]
> https://github.com/johannesjo/super-productivity/issues/2109#issuecomment-2945074611

@XanatosX, you mention devices, but I don't believe that hardware matters much here: the OneDrive APK for AOSP exists for both x86-64 and ARM64-V8A. Irrespective, it is possible, not least because Google Drive and NextCloud do the same:

1. On AOSP, [`com.microsoft.skydrive_7.32`](https://www.apkmirror.com/apk/microsoft-corporation/onedrive/microsoft-onedrive-7-32-beta-2-release/microsoft-onedrive-7-32-beta-2-android-apk-download/#wrap:~:text=(arm64-v8a)%20(Android%207.0+)-,Microsoft%20OneDrive%207.32%20(Beta%202),-(arm64-v8a)%20(Android%207.0+)) exposes its storage to `com.google.android.documentsui/com.android.documentsui.picker.PickActivity`:

   ![Image](https://github.com/user-attachments/assets/22594224-760b-4bb7-ab44-53d1785d5068)

1. On i(Pad)OS, [`apps.apple.com/us/app/microsoft-onedrive/id477537958`](https://apps.apple.com/us/app/microsoft-onedrive/id477537958) exposes its storage to `UIDocumentPickerViewController`:

   > ![Image](https://github.com/user-attachments/assets/6de95f79-4a75-49a1-b4c2-47defd45e350)

**Comment by XanatosX** [comment]
Ahh I think I did missunderstood your initial idea. If the data is really exposed already it would be nice if the current implementation for the local sync could be extended to store the data inside that specific folder letting the OneDrive app syncing the data.

As I do not use an IOS device I do not know if you can access the files directly from within another app.

**Comment by beedell-roke** [comment]
> As I do not use an IOS device I do not know if you can access the files directly from within another app.

@XanatosX, the "cloud" storage provider exposes its backend storage via a virtual filesystem API on AOSP and all Darwin-based OSes, so, as aforestated, they can, transparently to the consumer.

> If the data is really exposed already it would be nice if the current implementation for the local sync could be extended to store the data inside that specific folder letting the OneDrive app syncing the data.

Currently, can the location for the local database be chosen by the user? If not, that's indeed what should be remediated, for it requires no "cloud" API integration, and maintenance thereof.

**Comment by OLLI-S** [comment]
In the Windows app, I am able to select "Local File" as sync and then select the OneDrive folder
`C:\Users\MyUserName\OneDrive\SuperProductivity`

But how do I open this folder from the Android app?

**Comment by beedell-roke** [comment]
> https://github.com/johannesjo/super-productivity/issues/2109#issuecomment-2957067961

@OLLI-S, hopefully the undermentioned explains. In retrospect, it might be somewhat unnecessarily comprehensive, but more is usually better than less:

<table>
<thead>
<tr>
<th>OS</th>
<th>Location (PowerShell Core)</th>
</tr>
</thead>
<tbody><tr>
<td><code>cpe:/o:google:android:14</code></td>
<td>

For the actual <em>path</em>, you&#39;ll need to follow <a href="https://android.stackexchange.com/q/261627/311449"><code>android.stackexchange.com/q/261627</code></a> for an answer, not that it'll be of much use to you.

Otherwise, you need to locate a file *picker* [^2] which integrates with the <code>DocumentProvider</code> API. This is rare, [^1] so it's lucky (rather, expected) that the OS-included <code>com.android.documentsui.files.FilesActivity</code> (&quot;Files&quot;) does so:

https://github.com/user-attachments/assets/979275e2-22a1-417d-8a8e-896f913db2a1

<details>

<ol>
<li>

![Image](https://github.com/user-attachments/assets/7c79e4d8-cbb4-4e7c-9a83-489bbd28c4c9)

</li>
<li>

![Image](https://github.com/user-attachments/assets/38441b7b-4e2c-4333-8ea5-fa04220b24e0)

</li>
</ol>

</details>

If you want it to *synchronise* this way, that requires that Super Productivity invoke an `android.intent.action.OPEN_DOCUMENT` intent, and utilise the Storage Access Framework to retain a connection to that location, currently, it does not: it solely permits access for import and export.

</td>
</tr>
<tr>
<td><code>cpe:/o:apple:macos</code></td>
<td><code>[RegEx]&quot;\$Env:HOME\/Library\/CloudStorage\/OneDrive.*&quot;</code></td>
</tr>
<tr>
<td><code>cpe:/o:microsoft:windows_11_24h2</code></td>
<td><code>Start-Process "$Env:SystemDrive\Users\[Environment::Username]\OneDrive"</code></td>
</tr>
<tr>
<td><code>cpe:/o:apple:</code>&#39;s <code>ipados</code> and <code>iphone_os</code></td>
<td>For the path, follow <a href="https://apple.stackexchange.com/q/480308/403705"><code>apple.stackexchange.com/q/480308</code></a> for an answer, or read the documentation for <code>UIDocumentPickerViewController</code> (if it includes it).</td>
</tr>
</tbody></table>

[^1]: [`stackoverflow.com/revisions/72859302/1`](https://stackoverflow.com/revisions/72859302/1#content:~:text=In%20practice%2C%20I%20am%20not%20surprised%20that%20few%20file%20managers%20do%20this.)

[^2]: [`android.stackexchange.com/revisions/210168/2`](https://android.stackexchange.com/revisions/210168/2#:~:text=I%20can%20pick%20attachments%20with%20any%20File%20Manager%20that%20registers%20one%20of%20its%20activities%20as%20a%20preferred%20activity%20for%20some%20implicit%20intent%20action;%20android.intent.action.OPEN_DOCUMENT%20in%20our%20case.)

**Comment by OLLI-S** [comment]
@RokeJulianLockhart 
I was asking my question under user-perspective (how as a user I can store my files on OneDrive and also access it via the Android app).
Your answer seems to be for developers, so I should have posted my questions in the **Discussions** section.

**Comment by beedell-roke** [comment]
> https://github.com/johannesjo/super-productivity/issues/2109#issuecomment-2957849101

@OLLI-S, much of what I've stated applies to a mere "user" too: I'm no AOSP developer. As an example, have you observed the screencast? It depicts how to invoke the file picker from within SP's GUI, and navigate to the OneDrive virtual filesystem.

**Comment by jeslinmx** [comment]
> > As I do not use an IOS device I do not know if you can access the files directly from within another app.
> 
> [@XanatosX](https://github.com/XanatosX), the "cloud" storage provider exposes its backend storage via a virtual filesystem API on AOSP and all Darwin-based OSes, so, as aforestated, they can, transparently to the consumer.
> 
> > If the data is really exposed already it would be nice if the current implementation for the local sync could be extended to store the data inside that specific folder letting the OneDrive app syncing the data.
> 
> Currently, can the location for the local database be chosen by the user? If not, that's indeed what should be remediated, for it requires no "cloud" API integration, and maintenance thereof.

There is no native iOS super-productivity app, though, and saving to local storage on iOS is not available. Regardless, using the super-productivity -> local file system -> Onedrive app route is quite unreliable, as the syncing is decoupled from the reading and writing. Especially on iOS, where the sync scheduling is at the whim of the OS and not the app, so you may end up with all sorts of weird sync conflicts if you make changes from mobile, and make changes on the desktop end before the Onedrive is activated to do the syncing, which can be a few hours.

**Comment by boschet** [comment]
First of all, I would be very happy about OneDrive or Google support. Unfortunately I did not get it to work with the guide above.

**Comment by ArminWeb** [comment]
Same issues here!
Would be such a useful app, if only the OneDrive sync works properly.
Google is IMHO out of interest due to privacy concerns.

**Comment by OLLI-S** [comment]
Sorry, that I created a separate topic in "Discussions", but I thought, that this is an issue on my side (that I just need some help).
But this is really a feature request.

In the Windows app, I am able to select "Local File" as sync provider and then select the OneDrive folder
`C:\Users\MyUserName\OneDrive\SuperProductivity`
In this folder a file called `__meta_` is created.

On Android I can not select the OneDrive folder within Super Productivity.
When I try to import the file `C:\Users\MyUserName\OneDrive\SuperProductivity\__meta_` then I get an error, that this is not a valid file.

I do now want to get a hosting contract, just because of Super Productivity.
Therefore, please enhance the OneDrive support.

@johannesjo Maybe you offer a paid WebDAV for users?

**Comment by johannesjo** [comment]
> @johannesjo Maybe you offer a paid WebDAV for users?

Yes, this is maybe something we should do in the future to offer users an easy and fast way to use sync. I've to think about it in detail.


PRs for the one drive integration would be very welcome! Anyone up for the challenge?

**Comment by OLLI-S** [comment]
I think the only limitation regarding OneDrive is, that I can't select a OneDrive folder in Super Productivity.
Some file managers (like "X-Plore") allow to open OneDrive folders.
Here I need to connect to OneDrive inside X-Plore and OneDrive is asking for permissions (if X-Plore can access OneDrive).
Maybe there is an Open-Spurce file manager, that also allows to access OneDrive (where you get an idea, how to proceed)?

**Comment by boschet** [comment]
> I think the only limitation regarding OneDrive is, that I can't select a OneDrive folder in Super Productivity.
> Some file managers (like "X-Plore") allow to open OneDrive folders.
> Here I need to connect to OneDrive inside X-Plore and OneDrive is asking for permissions (if X-Plore can access OneDrive).
> Maybe there is an Open-Spurce file manager, that also allows to access OneDrive (where you get an idea, how to proceed)?

It's more difficult than that. The official OneDrive app does not sync the all the files to the local filesystem like the app on windows does. Third party file managers include the OneDrive API transparently to the gui user. But there is no file created in the file system. Therefore, the OneDrive API needs to be implemented or tools like folder sync can be used. Unfortunately not all cloud storage provider offer WebDAV access.

**Comment by OLLI-S** [comment]
@funnym0nk3y I did not find a way to use OneDrive via WebDAV

**Comment by OLLI-S** [comment]
I am using Super Productivity (the portable version) on my PC.
Here I have set-up a custom data folder on a **local hard drive** by using the command line parameter:
`--user-data-dir="D:\folder\subfolder"`

Today I am using my laptop, where I have not yet used Super Productivity before.
I have downloaded the app (the portable version) but now I have no data.

My plan is to store all data directly on OneDrive:
* Create a folder `SuperProductivity` in Onedrive
* Copy all data from the existing custom data folder (`D:\folder\subfolder`) to this new folder on OneDrive
* Start Super Productivity with the command line parameter `--user-data-dir="C:\Users\<MyUser>\OneDrive\SuperProductivity"`

Would this work or will there be any issues (if Super Productivity on my PC and on my Laptop have the same folder as custom data folder?
Or should I start Super Productivity on my laptop *without* any command line parameter and use the built-in sync feature (I have set the sync provider to **LocalFile** and have set the **Sync Foder Path** to `C:\Users\<MyUser>\OneDrive\SuperProductivity`).
